### DISCUSSION TITLE
Avalonia program UI is laggy

### DISCUSSION THREAD

**Discussion by rocinn** [AvaloniaUI/Avalonia#16858] [Q&A] [answered]
I'm encountering an issue with my Avalonia program that seems to be related to the Windows system's memory resource recycling. The phenomenon is as follows:
When my Avalonia program starts, the memory usage is generally between 100MB and 200MB. After running for a long time without any operation (for example, 12 hours), if I try to interact with the UI using the mouse, I find it very laggy (for example, clicking a button to pop up a window takes several seconds or even longer). However, when I click the same button for the second time, the response is very quick.
By observing the memory usage of the program, after a long period of inactivity, the program's memory usage drops to the lowest point, usually around 40MB. When interacting with the interface, the memory usage will continue to rise (after all buttons or interfaces have been operated, the memory usage will tend to stabilize, for example, fluctuating around 400MB).

I'm not quite clear about what the reason is, it looks like it's a very normal Windows memory page recycling.

Operating environment:
Windows 10
CPU: Celeron G5905 (3.5GHz, 2 cores, 2 threads)
Memory: 8GB
Hard drive: Mechanical hard drive

Additional tests, the same program: will not exhibit the above lag issues
Operating environment:
Windows 7
CPU: i5 4590 (3.3GHz, 4 cores, 4 threads)
Memory: 4GB
Hard drive: Mechanical hard drive

**Comment by jmacato** [comment, marked as answer]
Hi!

The lag you’re experiencing after long periods of inactivity is likely due to memory swapping to the hard drive. When your program is idle, Windows moves less-used data from RAM to the pagefile on the hard drive, causing delays when you interact with the UI again.

The difference in performance between your two setups is due to hardware differences, particularly the CPU and RAM. The Windows 10 system with a Celeron G5905 CPU and 8GB of RAM is more prone to paging compared to the Windows 7 system.

To mitigate this, consider:

- Increasing physical RAM.
- Optimizing pagefile settings.
- Upgrading to an SSD.

I hope this helps.

**Reply by rocinn** [reply]
Thank you very much for your explanation.

**Reply by rocinn** [reply]
Thank you very much, by turning off Windows virtual memory, the Avalonia UI no longer lags after long-term operation. 👍
